libraries/networking/src/ExternalResource.h

```diff
@@ -0,0 +1,139 @@
+        Content,
+
+        /** Assets that used to be in the mpassets S3 bucket (hifi marketplace) */
+        MPAssets,
+
+        /** Vircadia assets */
+        Assets
+    };
+
+    Q_ENUM(Bucket)
+
+    /**
+     * Returns the location of a resource as a QUrl
+     *
+     * Returns the location of the resource \p relative_path in bucket \p bucket
+     *
+     * @note The resulting path will be sanitized by condensing multiple instances of '/' to one.
+     * This is done for easier usage with Amazon S3 and compatible systems.
+     *
+     * @param bucket The bucket in which the resource is found
+     * @param relative_path The path of the resource within the bucket
+     * @returns The resulting URL as a QUrl
+     */
+    Q_INVOKABLE QUrl getQUrl(Bucket bucket, const QUrl& relative_path);
+
+
+    /**
+     * Returns the location of a resource as a QUrl
+     *
+     * Returns the location of the resource \p relative_path in bucket \p bucket
+     *
+     * @note The resulting path will be sanitized by condensing multiple instances of '/' to one.
+     * This is done for easier usage with Amazon S3 and compatible systems.
+     *
+     * @param bucket The bucket in which the resource is found
+     * @param relative_path The path of the resource within the bucket
+     * @returns The resulting URL as a QUrl
+     */
+    Q_INVOKABLE QUrl getQUrl(Bucket bucket, QString path) {
+        return getQUrl(bucket, QUrl(path));
+    }
+
+    /**
+     * Returns the location of a resource as a QString
+     *
+     * Returns the location of the resource \p relative_path in bucket \p bucket
+     *
+     * @note The resulting path will be sanitized by condensing multiple instances of '/' to one.
+     * This is done for easier usage with Amazon S3 and compatible systems.
+     *
+     * @param bucket The bucket in which the resource is found
+     * @param relative_path The path of the resource within the bucket
+     * @returns The resulting URL as a QString
+     */
+    Q_INVOKABLE QString getUrl(Bucket bucket, const QUrl &relative_path) {
+        return ExternalResource::getQUrl(bucket, relative_path).toString();
+    };
+
+    /**
+     * Returns the location of a resource as a QString
+     *
+     * Returns the location of the resource \p relative_path in bucket \p bucket
+     *
+     * @note The resulting path will be sanitized by condensing multiple instances of '/' to one.
+     * This is done for easier usage with Amazon S3 and compatible systems.
+     *
+     * @param bucket The bucket in which the resource is found
+     * @param relative_path The path of the resource within the bucket
+     * @returns The resulting URL as a QString
+     */
+    Q_INVOKABLE QString getUrl(Bucket bucket, const QString &relative_path) {
+        return ExternalResource::getQUrl(bucket, QUrl(relative_path)).toString();
+    };
+
+    private:
+    ExternalResource(QObject* parent = nullptr);
+
+    QMap<Bucket, QUrl> _bucketBases {
```

Comment:
nitpicking (it really doesn't matter for this small map), but what you want here is really more like just a list, especially since the values are hardcoded.  a QMap is a red-black tree with non-constant lookup, and a QHash is a proper hashmap with constant lookup but extra memory footprint.  also this could be const and static

Reply:
For now I've left this as-is, as I found a benchmark suggesting QMap is the faster choice for small numbers of items. Also hopefully the latest updates show better what I had in mind. Now the map can actually be modified.